microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/corrections/proposal/InsertAnnotationAttributeProposal.java

```diff
@@ -0,0 +1,137 @@
+import org.eclipse.lsp4j.CodeActionKind;
+
+public class InsertAnnotationAttributeProposal extends LinkedCorrectionProposal {
+
+	private final Annotation fAnnotation;
+	private final Set<String> attributes;
+
+	public InsertAnnotationAttributeProposal(String label, ICompilationUnit targetCU, Annotation annotation,
+			int relevance, String... attributes) {
+		super(label, CodeActionKind.QuickFix, targetCU, null, relevance);
+		this.fAnnotation = annotation;
+		this.attributes = new HashSet<>(Arrays.asList(attributes));
+	}
+
+	@Override
+	protected ASTRewrite getRewrite() throws CoreException {
+		AST ast = fAnnotation.getAST();
+
+		ASTRewrite rewrite = ASTRewrite.create(ast);
+		createImportRewrite((CompilationUnit) fAnnotation.getRoot());
+
+		ListRewrite listRewrite;
+		if (fAnnotation instanceof NormalAnnotation) {
+			listRewrite = rewrite.getListRewrite(fAnnotation, NormalAnnotation.VALUES_PROPERTY);
+		} else {
+			NormalAnnotation newAnnotation = ast.newNormalAnnotation();
+			newAnnotation.setTypeName((Name) rewrite.createMoveTarget(fAnnotation.getTypeName()));
+			rewrite.replace(fAnnotation, newAnnotation, null);
+
+			listRewrite = rewrite.getListRewrite(newAnnotation, NormalAnnotation.VALUES_PROPERTY);
+		}
+		addDefinedAtributes(fAnnotation.resolveTypeBinding(), listRewrite);
+
+		return rewrite;
+	}
+
+	private void addDefinedAtributes(ITypeBinding binding, ListRewrite listRewriter) {
+		Set<String> implementedAttribs = new HashSet<String>();
+		if (fAnnotation instanceof NormalAnnotation) {
+			List<MemberValuePair> list = ((NormalAnnotation) fAnnotation).values();
+			for (int i = 0; i < list.size(); i++) {
+				MemberValuePair curr = list.get(i);
+				implementedAttribs.add(curr.getName().getIdentifier());
+			}
+		} else if (fAnnotation instanceof SingleMemberAnnotation) {
+			implementedAttribs.add("value"); //$NON-NLS-1$
+		}
+		ASTRewrite rewriter = listRewriter.getASTRewrite();
+		AST ast = rewriter.getAST();
+		ImportRewriteContext context = null;
+		ASTNode bodyDeclaration = ASTResolving.findParentBodyDeclaration(listRewriter.getParent());
+		if (bodyDeclaration != null) {
+			context = new ContextSensitiveImportRewriteContext(bodyDeclaration, getImportRewrite());
+		}
+
+		IMethodBinding[] declaredMethods = binding.getDeclaredMethods();
+		for (int i = 0; i < declaredMethods.length; i++) {
+			IMethodBinding curr = declaredMethods[i];
+			if (!implementedAttribs.contains(curr.getName()) && attributes.contains(curr.getName())) {
+				MemberValuePair pair = ast.newMemberValuePair();
+				pair.setName(ast.newSimpleName(curr.getName()));
+				pair.setValue(newDefaultExpression(ast, curr.getReturnType(), context));
+				listRewriter.insertLast(pair, null);
+
+				addLinkedPosition(rewriter.track(pair.getName()), false, "val_name_" + i); //$NON-NLS-1$
+				addLinkedPosition(rewriter.track(pair.getValue()), false, "val_type_" + i); //$NON-NLS-1$
```

Comment:
Just a note about the addLinkedPosition(..) calls. I believe these aren't needed, but I'm fine with keeping them in for potential future enhancements.

In case their use is not clear, linked positions in Eclipse are equivalent to snippet text in VS Code. So when you apply some piece of text and you have some placeholders that you want a user to quickly cycle through (eg. with tab) and modify, this is what is used on the Eclipse side. Notice that with your patch, when you apply the member value insertion, there's no way to cycle between those elements. I think code action doesn't support snippet text the same way completion does, so we can't use this.

However, in the future, if code actions adopt snippet text in the LSP, we could potentially, get the new positions from the underlying proposal and generate a code action that supports snippets (in ChangeUtil).

Reply:
Thanks for the good  explanation (I have already linked psoition). You are right we should remove those lines code, it's a copy/paste problem.

I remove those line code.